src/v/pandaproxy/schema_registry/sharded_store.cc

```diff
@@ -227,15 +227,66 @@ ss::future<bool> sharded_store::upsert_subject(
       deleted);
 }
 
-ss::future<schema_id> sharded_store::allocate_schema_id() {
-    auto increment = [this] { return _next_schema_id++; };
+/// \brief Get the schema ID to be used for next insert
+ss::future<schema_id> sharded_store::project_schema_id() {
+    // This is very simple because we only allow one write in
+    // flight at a time.  Could be extended to track N in flight
+    // operations if needed.  _next_schema_id gets updated
+    // if the operation was successful, as a side effect
+    // of applying the write to the store.
+    auto fetch = [this] { return _next_schema_id; };
     co_return co_await ss::smp::submit_to(
+      ss::shard_id{0}, _smp_opts, std::move(fetch));
+}
+
+ss::future<foreign_sem_units> sharded_store::begin_write() {
+    auto remote_fn = [this]() -> ss::future<foreign_sem_units> {
+        auto units = ss::make_foreign(
+          std::make_unique<sem_units>(co_await ss::get_units(_seq_sem, 1)));
+
+        co_return units;
```

Comment:
This a global lock of sorts, I agree it's a red flag. Because there are multiple shard keys in play here (schema_id and subject), I created sharded_store to encapsulate the trickiness. It's the right place to do it, but I think we should look to improve this soon. Just going to mention #1869, which has a similar problem (requires single core/global dispatching), but with a narrower scope.

Reply:
I think it is supposed to guarantee that, but I'm revising this to go the `peering_sharded_service` route and have things that require the semaphores `invoke_on` the core instead, so no more foreign pointers.